src/fast_wilcox.cpp

```diff
@@ -0,0 +1,232 @@
+                    sortspace.emplace_back(exprs_x[j], exprs_i[j]);
+                }
+            }
+            std::sort(sortspace.begin(), sortspace.end());
+            last = target_id[i];
+        }
+
+        // Identifying the cells with TF expression + peak.
+        {
+            int tf_offset = exprs_p[tf_id[i]], tf_last = exprs_p[tf_id[i]+1];
+            int peak_offset = exprs_p[peak_id[i]], peak_last = exprs_p[peak_id[i]+1];
+            int k = peak_offset;
+
+            for (int j = tf_offset; j < tf_last; ++j) {
+                if (exprs_x[j] <= 0) {
+                    continue;
+                }
+
+                int icurrent = exprs_i[j];
+                while (k < peak_last && peak_i[k] < icurrent) {
+                    ++k;
+                }
+
+                if (k == peak_last) {
+                    break;
+                }
+
+                if (peak_i[k] == icurrent && peak_x[k] > 0) {
+                    okay_indices.push_back(icurrent);
+                    okay[icurrent] = 1;
+                }
+            }
+        }
+
+        {
+            std::fill(okay_zeros.begin(), okay_zeros.end(), 0);
+            std::fill(okay_total.begin(), okay_total.end(), 0);
+            std::fill(notokay_zeros.begin(), notokay_zeros.end(), 0);
+            std::fill(notokay_total.begin(), notokay_total.end(), 0);
+
+            // Looping through the sortspace and creating cluster-specific sorted vectors.
+            for (const auto& ss : sortspace) {
+                int c = clusters[ss.second];
+                if (okay[ss.second]) {
+                    ++okay_zeros[c];
+                } else {
+                    ++notokay_zeros[c];
+                }
+            }
+
+            for (auto x : okay_indices) {
+                ++okay_total[clusters[x]];
+            }
+
+            // Computing the U statistic between the cells with TF+peak, and those without both.
+            for (int c = 0; c < ngroups; ++c) {
+                okay_zeros[c] = okay_total[c] - okay_zeros[c];
+                notokay_total[c] = cluster_sizes[c] - okay_total[c];
+                notokay_zeros[c] = notokay_total[c] - notokay_zeros[c];
+            }
+
+            auto col_u = output_auc.column(i);
+            auto output_auc_ptr = static_cast<double*>(col_u.begin());
+            auto col_tie = output_ties.column(i);
+            auto output_tie_ptr = static_cast<double*>(col_tie.begin());
+            compute_auc(sortspace, notokay_zeros, okay_zeros, clusters.begin(), okay.data(), workspace, output_auc_ptr, output_tie_ptr);
+
+            full_okay_zeros[0] = std::accumulate(okay_zeros.begin(), okay_zeros.end(), 0);
+            full_notokay_zeros[0] = std::accumulate(notokay_zeros.begin(), notokay_zeros.end(), 0);
+            compute_auc(sortspace, full_okay_zeros, full_notokay_zeros, full_cluster.data(), okay.data(), full_workspace, output_auc_ptr + ngroups, output_tie_ptr + ngroups);
```

Comment:
The positions of arguments `full_okay_zeros` and `full_notokay_zeros` should be interchanged.

Reply:
:+1: